### DISCUSSION TITLE
Dynamic Addition of Query Runners in Redash Without Rebuilding (Beginner Question)

### DISCUSSION THREAD

**Discussion by johnthotekat** [getredash/redash#7183] [Q&A]
I'm new to Redash and still learning about its architecture and capabilities. As I explore its features, I'm curious about the possibility of extending its functionality by adding new query runners dynamically.

From what I understand so far, adding a new query runner typically requires modifying the core codebase and rebuilding the application. However, I'm wondering if there's a more flexible approach.
My question is: Is there a way for a Redash beginner to add new query runners dynamically at runtime, without having to rebuild and redeploy the entire application?

Specifically, I'd like to know:

1. Are there any existing user-friendly mechanisms or plugins in Redash that allow for dynamic addition of query runners?
2. If not, what would be the recommended approach for a Redash newcomer to implement or request such functionality?
3. What are the potential challenges or security considerations that a beginner should be aware of when thinking about this kind of feature?

As someone new to Redash, I'm open to learning about any existing solutions, workarounds, or even alternative approaches that might address this need. Any beginner-friendly insights or explanations would be greatly appreciated.

**Comment by zachliu** [comment]
Query runners are essentially python code that sends raw query texts to a database. Each database has some unique aspects to which the python code needs to adapt. Following best practices, Redash maintainers and contributors should thoroughly test their runner code. 

From a security perspective, all accesses to the db should be managed upon the creation of the db users. AFAIK, Redash has little way or responsibility to prevent a `DROP ALL DATABASES` from being executed. Some Redash users do use it for good with admin privileges.

**Reply by johnthotekat** [reply]
Thanks @zachliu for the context. I understand that the query runners are python code.  Which is placed up here : [https://github.com/getredash/redash/tree/master/redash/query_runner](url). And I was thinking of writing a custom query runner, which would be inhertiting BaseHTTPQueryRunner. To write and test the query runner, building redash again and deploying becomes a productivity killer. I'm keenly interested to know if we've some way to dynamically link the query runner custom class to redash.

**Reply by zachliu** [reply]
Ah, I thought you meant dynamically adding runners in a production environment, which requires a fundamental change on Redash.

Follow [this guide](https://github.com/getredash/redash/wiki/Local-development-setup), your dev environment will be able to load whatever code changes dynamically.

**Reply by abrie-stubber** [reply]
This seems strange, it would be incredibly useful (and simple) to be able to dynamically specify query_runners. This can easily be done and solve 90% of users' issues by simply updating the docker version to "accept" a volume where it can check for dynamic query_runners, and simply copy them in. The onus would be on the users to correctly create their runners and add them to the volume, which is perfect. 

Then there is the case of the `URL` query_runner ([here](https://github.com/getredash/redash/blob/master/redash/query_runner/url.py)). This query_runner would also solve the problem of anyone who is trying to dynamically add some query_runner to get some data into Redash, because they can set up an API that would return the data in the correct format. However, it is not visible on the front end, even if you have `REDASH_ADDITIONAL_QUERY_RUNNERS: "redash.query_runner.url"`. I believe it's because it doesnt have a `configuration_schema`. Why can't we use this ridiculously useful runner without custom building Redash?

How are users supposed to build dashboards that rely on some data that is only available in eg. Redis, when the entire Redash stack has to be custom built and reconfigured. This is not something that can simply be done in most businesses.
